cli/azd/pkg/async/task_context.go

```diff
@@ -0,0 +1,97 @@
+//nolint:all
+package async
+
+// Task function definition
+type TaskRunFunc[R comparable] func(taskContext *TaskContext[R])
+
+// The context available to the executing Task
+type TaskContext[R comparable] struct {
+	task   *Task[R]
+	error  error
+	result R
+}
+
+// Creates a new Task context
+func NewTaskContext[R comparable](task *Task[R]) *TaskContext[R] {
+	return &TaskContext[R]{
+		task: task,
+	}
+}
+
+// Sets the specified error for the task
+func (c *TaskContext[R]) SetError(err error) {
+	if c.result != *new(R) {
+		panic("Task result has already been set! Task cannot have both a result and an error.")
+	}
+
+	if c.error != nil {
+		panic("Task error has already been set! Ensure your task error is only ever set one time.")
+	}
+
+	c.error = err
+}
+
+// Sets the result of the Task
+func (c *TaskContext[R]) SetResult(result R) {
+	if c.error != nil {
+		panic("Task error has already been set! Task cannot have both a result and an error.")
+	}
+
+	if c.result != *new(R) {
+		panic("Task result has already been set! Ensure your task result is only ever set one time.")
+	}
+
+	c.result = result
+}
+
+// Task with progress function definition
+type TaskWithProgressRunFunc[R comparable, P comparable] func(ctx *TaskContextWithProgress[R, P])
+
+// The context available to the executing Task
+type TaskContextWithProgress[R comparable, P comparable] struct {
+	task *TaskWithProgress[R, P]
+	TaskContext[R]
+}
+
+// Creates a new Task context with progress reporting
+func NewTaskContextWithProgress[R comparable, P comparable](task *TaskWithProgress[R, P]) *TaskContextWithProgress[R, P] {
+	innerTask := NewTaskContext(&task.Task)
+
+	return &TaskContextWithProgress[R, P]{
+		task:        task,
+		TaskContext: *innerTask,
+	}
+}
+
+// Write a new progress value to the underlying progress channel
+func (c *TaskContextWithProgress[R, P]) SetProgress(progress P) {
+	c.task.progressChannel <- progress
+}
+
+// Task with progress function definition
+type InteractiveTaskWithProgressRunFunc[R comparable, P comparable] func(ctx *InteractiveTaskContextWithProgress[R, P])
```

Comment:
Here I would consider de-coupling the `context` parameter from the facility to indicate that the interactive portion of the task starts or ends (and pass the latter as an extra parameter to the task function). This way the context parameter does not have to be special in any way.

Reply:
Good feedback, but I'm going to keep this as is for the moment and see how it feels.  If later this is too cumbersome I am open to refactoring.